### PR TITLE
Low latency cryptocurrency information access via websockets.

### PR DESCRIPTION
**Pull request by Alescontrela**
Addresses #14 by extending `CryptoBook`'s functionality to allow for low-latency data retrieval from CryptoCompare. To this aim, this PR proposes the following changes:

* The introduction of `CryptoBook-js`, which contains a javascript web-socket that retrieves data from the `CryptoCompare` streaming API and publishes the data to subscribed clients.
  * Tests for this methods utility functions are included.
* A new directory that separates the codebase into `CryptoBook-py` and `CryptoBook-js`, such that new modules are stored in the directory of the prevalent programming language of the new module. 
* A `docker-compose.yml` file is introduced to launch two separate docker containers: one for historical data retrieval (`CryptoBook-py`), and another for real-time data retrieval (`CryptoBook-js`).

### CODE REVIEW

**Comment by synchronizing** [comment]
> ## (aescontrela/synchronizing): make all these calls asynchronously
> https://github.com/Waultics/CryptoBook/blob/938162a15f843432b5dc88eae73d8a53ae15f6be/CryptoBook/utils.py#L74-L77
> 
> ###### This comment was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in [938162a](https://github.com/Waultics/CryptoBook/commit/938162a15f843432b5dc88eae73d8a53ae15f6be) in #15. cc @Waultics.

To be able to successfully achieve this, we can do either of two things:

1. Guesstimate the end time of each call, and then organize the data. 
2. Do the initial call, and use it as a bases to the length of time that the data is returned in to generate the lengths that need to be called asynchronously to fill the date throughout. Check the data is sequential, and then return.

**Comment by Alescontrela** [comment]
Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing

**Comment by synchronizing** [comment]
> Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing

100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the Dockerfile for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.

**Comment by synchronizing** [comment]
> > ## (aescontrela/synchronizing): make all these calls asynchronously
> > https://github.com/Waultics/CryptoBook/blob/938162a15f843432b5dc88eae73d8a53ae15f6be/CryptoBook/utils.py#L74-L77
> > 
> > ###### This comment was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in [938162a](https://github.com/Waultics/CryptoBook/commit/938162a15f843432b5dc88eae73d8a53ae15f6be) in #15. cc @Waultics.
> 
> To be able to successfully achieve this, we can do either of two things:
> 
> 1. Guesstimate the end time of each call, and then organize the data.
> 2. Do the initial call, and use it as a bases to the length of time that the data is returned in to generate the lengths that need to be called asynchronously to fill the date throughout. Check the data is sequential, and then return.
> 
> I believe number 2 is the best option, as I think its fairly safe to assume that the exchanges will always return a constant amount of rows per call.

Fixed within #16.

**Comment by Alescontrela** [comment]
> > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> 
> 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.

I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these

**Comment by synchronizing** [comment]
> > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > 
> > 
> > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> 
> I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these

Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.

**Comment by Alescontrela** [comment]
> > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > 
> > > 
> > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > 
> > 
> > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> 
> Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.

Yep, agreed. Let's just open an issue for it and close once resolved.

**Comment by synchronizing** [comment]
> > > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > > 
> > > > 
> > > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > > 
> > > 
> > > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> > 
> > 
> > Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.
> 
> Yep, agreed. Let's just open an issue for it and close once resolved.

Sounds good!

**Comment by Alescontrela** [comment]
> > > > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > > > 
> > > > > 
> > > > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > > > 
> > > > 
> > > > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> > > 
> > > 
> > > Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.
> > 
> > 
> > Yep, agreed. Let's just open an issue for it and close once resolved.
> 
> Sounds good!

#24 ...


let's see how deep we can make this thread 😎

**Comment by synchronizing** [comment]
> > > > > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > > > > 
> > > > > > 
> > > > > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > > > > 
> > > > > 
> > > > > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> > > > 
> > > > 
> > > > Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.
> > > 
> > > 
> > > Yep, agreed. Let's just open an issue for it and close once resolved.
> > 
> > 
> > Sounds good!
> 
> #24 ...
> 
> let's see how deep we can make this thread 😎

#24 looking good 🤙🏼

**Comment by Alescontrela** [comment]
> > > > > > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > > > > > 
> > > > > > > 
> > > > > > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > > > > > 
> > > > > > 
> > > > > > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> > > > > 
> > > > > 
> > > > > Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.
> > > > 
> > > > 
> > > > Yep, agreed. Let's just open an issue for it and close once resolved.
> > > 
> > > 
> > > Sounds good!
> > 
> > 
> > #24 ...
> > let's see how deep we can make this thread 😎
> 
> #24 looking good 🤙🏼

Ok let's stop polluting this PR

**Comment by synchronizing** [comment]
> > > > > > > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > > > > > > 
> > > > > > > > 
> > > > > > > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > > > > > > 
> > > > > > > 
> > > > > > > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> > > > > > 
> > > > > > 
> > > > > > Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.
> > > > > 
> > > > > 
> > > > > Yep, agreed. Let's just open an issue for it and close once resolved.
> > > > 
> > > > 
> > > > Sounds good!
> > > 
> > > 
> > > #24 ...
> > > let's see how deep we can make this thread 😎
> > 
> > 
> > #24 looking good 🤙🏼
> 
> Ok let's stop polluting this PR

K.

**Comment by Alescontrela** [comment]
> > > > > > > > > > Moving on to create Docker container in which to run the javascript websocket server. Considering putting some example python clients in examples/js/*_example.py. Thoughts? @synchronizing
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > 100% agree -- easier to keep manageable. Feel free to move the current files in `examples/` to `examples/py/`. I believe we should also rename`CryptoBook/CryptoBook` folder into `CryptoBook/CryptoBook-py`, move over (and fix) the `Dockerfile` for that specific `-py` folder, and create a new `docker-compose.yml` file in the root. Lmk your thoughts on this.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > I think this is a good idea. I can see a future where we implement more js websockets for real-time data gathering and a CryptoBook-js directory would be the perfect place for these
> > > > > > > 
> > > > > > > 
> > > > > > > Awesome sounds like a plan. I would recommend doing that after the PR, since otherwise, we would be pushing old `py` code into the codebase.
> > > > > > 
> > > > > > 
> > > > > > Yep, agreed. Let's just open an issue for it and close once resolved.
> > > > > 
> > > > > 
> > > > > Sounds good!
> > > > 
> > > > 
> > > > #24 ...
> > > > let's see how deep we can make this thread 😎
> > > 
> > > 
> > > #24 looking good 🤙🏼
> > 
> > 
> > Ok let's stop polluting this PR
> 
> K.

Thank you 💯

File: .travis.yml [hunk 1]
```diff
@@ -1,15 +1,26 @@
-dist: xenial
-language: python
+matrix:
```
Comment (synchronizing) [inline .travis.yml]: Good work on this my dude.

File: CryptoBook-py/tests/test_api.py [hunk 2]
```diff
@@ -2,6 +2,8 @@
 import pytest
 import json
 
+import sys
+print(sys.path)
```
Comment (synchronizing) [inline CryptoBook-py/tests/test_api.py]: Perhaps remove the print?

**Comment by synchronizing** [comment]
Looks good to me. Passing Travis and moves project pretty well.

**Pull request merged**